modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java

```diff
@@ -386,6 +386,96 @@ private MapStream(Stream<Map.Entry<K, V>> underlying) {
     return underlying.flatMap(e -> mapper.apply(e.getKey(), e.getValue()));
   }
 
+  //-----------------------------------------------------------------------
+  /**
+   * Sorts the entries in the stream by comparing the keys using their natural ordering.
+   * <p>
+   * If the keys in this map stream are not {@code Comparable} a {@code java.lang.ClassCastException} may be thrown.
+   * In this case use {@link #sortedKeys(Comparator)} instead.
+   *
+   * @return the sorted stream
+   */
+  @SuppressWarnings("unchecked")
+  public MapStream<K, V> sortedKeys() {
+    Comparator<K> comparator = (Comparator<K>) Comparator.naturalOrder();
+    return wrap(underlying.sorted((e1, e2) -> comparator.compare(e1.getKey(), e2.getKey())));
+  }
+
+  /**
+   * Sorts the entries in the stream by comparing the keys using the supplied comparator.
+   *
+   * @param comparator  a comparator of keys
+   * @return the sorted stream
+   */
+  public MapStream<K, V> sortedKeys(Comparator<? super K> comparator) {
+    return wrap(underlying.sorted((e1, e2) -> comparator.compare(e1.getKey(), e2.getKey())));
+  }
+
+  /**
+   * Sorts the entries in the stream by comparing the values using their natural ordering.
+   * <p>
+   * If the values in this map stream are not {@code Comparable} a {@code java.lang.ClassCastException} may be thrown.
+   * In this case use {@link #sortedValues(Comparator)} instead.
+   *
+   * @return the sorted stream
+   */
+  @SuppressWarnings("unchecked")
+  public MapStream<K, V> sortedValues() {
+    Comparator<V> comparator = (Comparator<V>) Comparator.naturalOrder();
+    return wrap(underlying.sorted((e1, e2) -> comparator.compare(e1.getValue(), e2.getValue())));
+  }
+
+  /**
+   * Sorts the entries in the stream by comparing the values using the supplied comparator.
+   *
+   * @param comparator  a comparator of values
+   * @return the sorted stream
+   */
+  public MapStream<K, V> sortedValues(Comparator<? super V> comparator) {
+    return wrap(underlying.sorted((e1, e2) -> comparator.compare(e1.getValue(), e2.getValue())));
+  }
+
+  //-----------------------------------------------------------------------
+  /**
+   * Finds the minimum entry in the stream by comparing the keys using the supplied comparator.
+   *
+   * @param comparator  a comparator of keys
+   * @return the minimum entry
+   */
+  public Optional<Map.Entry<K, V>> minKeys(Comparator<? super K> comparator) {
```

Comment:
Should this be called `minKey` (singular)?

Reply:
I was thinking it would be minKeys because you're taking the minimum by comparing the keys, whereas minKey makes me think that it will return the minimum key i.e Optional<K>

Happy to be wrong.

Reply:
I think the name is fine as is, I just struggle to think that these four min/max methods would actually be used.

Reply:
Ah, I wanted to use them, that's why I thought of them.